File: internal/sim/batch.go
// Batch 是并发批量评估 API：多场景 × 多策略并行跑 sim 闭环，
// 结果确定性（每实例独立 Engine/Planner，无共享状态）。
// 使用场景：参数搜索（optsearch）、策略赛马、回归黄金集。
package sim

import (
	"fmt"
	"runtime"
	"sort"
	"sync"

	"github.com/deliciousbuding/arena/internal/domain"
	"github.com/deliciousbuding/arena/internal/strategy"
)

// BatchResult 是单次评估结果。
type BatchResult struct {
	Scene    string // 场景名
	Policy   string // 策略名
	Ticks    int
	Stats    SettleStats // 累计统计
	Final    *domain.TickState
	Score    float64 // 调用方评分函数结果（Evaluate 为空时 = 0）
	Timeline []TimelinePoint
}

// TimelinePoint 是逐 tick 关键指标采样（黄金集/赛马对比用）。
type TimelinePoint struct {
	Tick          int
	Resources     int
	ResourceCells int // 视野内可见资源格数（经济冻结诊断：reveal 不足）
	Workers       int
	Deposits      int // 累计 deposit 数（冻结检测：连续采样段不变 = 死锁）
	Kills         int
	UnitsLost     int
	Mode          string
}

// Evaluate 是评分函数（nil = 不评分）。输入为逐 tick 统计快照。
type Evaluate func(stats []SettleStats, final *domain.TickState) float64

// BatchOption 是批量评估配置。
type BatchOption struct {
	Workers  int // 并发数（<=0 = GOMAXPROCS）
	Interval int // Timeline 采样间隔（<=0 = 25）
	Evaluate Evaluate
}

// Batch 并发评估：scenes[i] × policies[j] 全组合并行跑 ticks 闭环。
// 高并发设计（Go 天然优势）：
//   - 固定 worker pool（不每任务起 goroutine）：goroutine 复用，
//     任务经 channel 分发，worker 数 = NumCPU（超卖配置）
//   - 无锁结果收集：预分配结果槽，每 worker 独占写入自己的槽
//     （无共享 mutex——共享锁在 8+ 并发下是扩展性瓶颈）
//   - 每 worker 独立 Engine/Planner/Refill/状态：零共享可变状态，
//     纯 embarrassingly parallel
//
// 结果确定性：同输入同输出（每组合独立实例，互不干扰）。
func Batch(scenes []*Scenario, policies []*strategy.Config, ticks int, opt BatchOption) []BatchResult {
	if opt.Workers <= 0 {
		opt.Workers = defaultWorkers()
	}
	if opt.Interval <= 0 {
		opt.Interval = 25
	}
	if len(scenes) == 0 || len(policies) == 0 {
		return nil
	}

	// 任务队列：预生成全组合索引（确定性顺序：scene 外循环 × policy 内循环）。
	type task struct {
		sceneIndex  int
		policyIndex int
		slotIndex   int
	}
	tasks := make([]task, 0, len(scenes)*len(policies))
	for sceneIndex := range scenes {
		for policyIndex := range policies {
			tasks = append(tasks, task{sceneIndex: sceneIndex, policyIndex: policyIndex, slotIndex: len(tasks)})
		}
	}

	// 无锁结果槽：每个任务一个独立槽，worker 只写自己的 slotIndex。
	slots := make([]BatchResult, len(tasks))
	workerCount := min(opt.Workers, len(tasks))
	taskChannel := make(chan task)
	var wg sync.WaitGroup
	for w := 0; w < workerCount; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for t := range taskChannel {
				slots[t.slotIndex] = runScenario(scenes[t.sceneIndex], policies[t.policyIndex], ticks, opt)
			}
		}()
	}
	for _, t := range tasks {
		taskChannel <- t
	}
	close(taskChannel)
	wg.Wait()

	// 确定性顺序：scene 名升序 × policy 名升序。
	sort.Slice(slots, func(i, j int) bool {
		if slots[i].Scene != slots[j].Scene {
			return slots[i].Scene < slots[j].Scene
		}
		return slots[i].Policy < slots[j].Policy
	})
	return slots
}

// runScenario 跑单个 场景×策略 组合（独立实例，可并发）。
// 热路径：每 tick 用 SettleInPlace 原地结算（不克隆——初始 CloneState
// 已保证 worker 间隔离；克隆是批量评估的 GC 主瓶颈）。
func runScenario(scene *Scenario, policy *strategy.Config, ticks int, opt BatchOption) BatchResult {
	state := scene.CloneState()
	planner := strategy.NewPlanner(*policy)
	engine := NewEngine()
	engine.Refill = NewRefillConfig(scene.LatentResources)

	result := BatchResult{Scene: scene.Name, Policy: policyName(policy), Ticks: ticks}
	statsPerTick := make([]SettleStats, 0, ticks)
	for tick := 1; tick <= ticks; tick++ {
		state.Tick = tick
		plan := planner.Decide(state)
		settled := engine.SettleInPlace(state, plan)
		result.Stats.Moves += settled.Stats.Moves
		result.Stats.Blocked += settled.Stats.Blocked
		result.Stats.Harvests += settled.Stats.Harvests
		result.Stats.Deposits += settled.Stats.Deposits
		result.Stats.Spawns += settled.Stats.Spawns
		result.Stats.SpawnBlocked += settled.Stats.SpawnBlocked
		result.Stats.ResourceDelta += settled.Stats.ResourceDelta
		result.Stats.Kills += settled.Stats.Kills
		result.Stats.ShotsFired += settled.Stats.ShotsFired
		result.Stats.SweepsFired += settled.Stats.SweepsFired
		result.Stats.UnitsLost += settled.Stats.UnitsLost
		result.Stats.HPRecovered += settled.Stats.HPRecovered
		result.Stats.ShieldRepaired += settled.Stats.ShieldRepaired
		statsPerTick = append(statsPerTick, settled.Stats)
		if tick%opt.Interval == 0 || tick == ticks {
			result.Timeline = append(result.Timeline, TimelinePoint{
				Tick:          tick,
				Resources:     state.Resources,
				ResourceCells: len(state.ResourceCells),
				Workers:       len(state.Workers),
				Deposits:      result.Stats.Deposits,
				Kills:         result.Stats.Kills,
				UnitsLost:     result.Stats.UnitsLost,
				Mode:          string(planner.DirectiveMode()),
			})
		}
	}
	result.Final = state
	if opt.Evaluate != nil {
		result.Score = opt.Evaluate(statsPerTick, state)
	}
	return result
}

// policyName 返回策略的可读名：优先 Config.Name（策略文件命名），
// 否则确定性字段拼接（默认策略/内联策略）。
func policyName(policy *strategy.Config) string {
	if policy.Name != "" {
		return policy.Name
	}
	return fmt.Sprintf("wt%d_r%d_er%d_pc%d_m%d",
		policy.WorkerTarget, policy.SpawnReserve, policy.ExploreRadius,
		policy.PopulationCeiling, policy.MilitaryRatio)
}

// Scenario 是独立场景（不依赖 game 包）：
// 初始状态 + 潜在资源池 + 场景名。CloneState 深拷贝供并发安全。
type Scenario struct {
	Name            string
	Initial         *domain.TickState
	LatentResources []domain.Position
}

// CloneState 深拷贝初始状态（并发安全：每 worker 独立副本）。
func (s *Scenario) CloneState() *domain.TickState {
	return cloneState(s.Initial)
}

// defaultWorkers 返回默认并发数 = NumCPU（超卖 1 倍：纯 CPU 批处理，
// 每 worker 独占核心；28 核机器 → 28 worker，无需再乘 2——评估间无
// I/O 等待，超卖只会增加调度抖动）。
func defaultWorkers() int {
	return runtime.NumCPU()
}

File: sim-rs/crates/cli/src/batch.rs
//! 并发批量评估（与 Go `internal/sim/batch.go` 对偶）。
//!
//! 多场景 × 多策略全组合 rayon 并行；结果确定性排序：scene 名升序 ×
//! policy 名升序（与 Go 一致）。每评估独立 Planner/Engine/Refill，
//! 零共享可变状态。

use arena_sim_domain::{Plan, Position, TickState};
use arena_sim_engine::{Engine, RefillConfig, SettleStats};
use arena_sim_strategy::{commander::DirectiveMode, Config, Planner};
use rayon::prelude::*;

use crate::policy_name::policy_name;

/// 独立场景（初始状态 + 潜在资源池）。
#[derive(Debug, Clone)]
pub struct Scenario {
    pub name: String,
    pub initial: TickState,
    pub latent_resources: Vec<Position>,
}

impl Scenario {
    /// 深拷贝初始状态（并发安全：每 worker 独立副本）。
    pub fn clone_state(&self) -> TickState {
        self.initial.clone()
    }
}

/// 逐 tick 关键指标采样（黄金集/赛马对比用；Interval 默认 25）。
#[derive(Debug, Clone)]
pub struct TimelinePoint {
    pub tick: i32,
    pub resources: i32,
    pub resource_cells: usize,
    pub workers: usize,
    pub deposits: i32,
    pub kills: i32,
    pub units_lost: i32,
    pub mode: String,
}

/// 单次评估结果（与 Go `BatchResult` 对偶）。
#[derive(Debug, Clone)]
pub struct BatchResult {
    pub scene: String,
    pub policy: String,
    pub ticks: i32,
    pub stats: SettleStats,
    pub final_state: TickState,
    pub score: f64,
    pub timeline: Vec<TimelinePoint>,
}

/// 批量评估配置。
#[derive(Debug, Clone)]
pub struct BatchOption {
    /// 并发线程数（0 = 默认 rayon 线程池）。
    pub workers: usize,
    /// Timeline 采样间隔（0 = 25）。
    pub interval: i32,
}

impl Default for BatchOption {
    fn default() -> Self {
        BatchOption {
            workers: 0,
            interval: 25,
        }
    }
}

/// 并发评估：scenes[i] × policies[j] 全组合并行跑 ticks 闭环。
/// 结果确定性顺序：scene 名升序 × policy 名升序。
pub fn batch(
    scenes: &[Scenario],
    policies: &[Config],
    ticks: i32,
    opt: BatchOption,
) -> Vec<BatchResult> {
    let interval = if opt.interval <= 0 { 25 } else { opt.interval };
    let thread_pool = if opt.workers > 0 {
        Some(
            rayon::ThreadPoolBuilder::new()
                .num_threads(opt.workers)
                .build()
                .expect("build rayon pool"),
        )
    } else {
        None
    };

    // 全组合索引（确定性顺序：scene 外循环 × policy 内循环）。
    let tasks: Vec<(usize, usize)> = (0..scenes.len())
        .flat_map(|scene_index| {
            (0..policies.len()).map(move |policy_index| (scene_index, policy_index))
        })
        .collect();

    let mut results: Vec<BatchResult> = if let Some(pool) = &thread_pool {
        pool.install(|| {
            tasks
                .par_iter()
                .map(|&(s, p)| run_scenario(&scenes[s], &policies[p], ticks, interval))
                .collect()
        })
    } else {
        tasks
            .par_iter()
            .map(|&(s, p)| run_scenario(&scenes[s], &policies[p], ticks, interval))
            .collect()
    };

    // 确定性顺序：scene 名升序 × policy 名升序。
    results.sort_by(|a, b| a.scene.cmp(&b.scene).then_with(|| a.policy.cmp(&b.policy)));
    results
}

/// 跑单个 场景×策略 组合（独立实例，可并发）。
fn run_scenario(scene: &Scenario, policy: &Config, ticks: i32, interval: i32) -> BatchResult {
    let mut state = scene.clone_state();
    let mut planner = Planner::new(policy.clone());
    let mut engine = Engine::new();
    engine.refill = Some(RefillConfig::new(&scene.latent_resources));

    let mut result = BatchResult {
        scene: scene.name.clone(),
        policy: policy_name(policy),
        ticks,
        stats: SettleStats::default(),
        final_state: state.clone(),
        score: 0.0,
        timeline: Vec::new(),
    };
    for tick in 1..=ticks {
        state.tick = tick;
        let plan: Plan = planner.decide(&state);
        let (_events, stats) = engine.settle_in_place(&mut state, &plan);
        result.stats.moves += stats.moves;
        result.stats.blocked += stats.blocked;
        result.stats.harvests += stats.harvests;
        result.stats.deposits += stats.deposits;
        result.stats.spawns += stats.spawns;
        result.stats.spawn_blocked += stats.spawn_blocked;
        result.stats.resource_delta += stats.resource_delta;
        result.stats.kills += stats.kills;
        result.stats.shots_fired += stats.shots_fired;
        result.stats.sweeps_fired += stats.sweeps_fired;
        result.stats.units_lost += stats.units_lost;
        result.stats.hp_recovered += stats.hp_recovered;
        result.stats.shield_repaired += stats.shield_repaired;
        if tick % interval == 0 || tick == ticks {
            result.timeline.push(TimelinePoint {
                tick,
                resources: state.resources,
                resource_cells: state.resource_cells.len(),
                workers: state.workers.len(),
                deposits: result.stats.deposits,
                kills: result.stats.kills,
                units_lost: result.stats.units_lost,
                mode: directive_mode_str(planner.directive_mode()),
            });
        }
    }
    result.final_state = state;
    result
}

/// 指挥模式字符串（与 Go DirectiveMode 一致）。
pub fn directive_mode_str(mode: DirectiveMode) -> String {
    match mode {
        DirectiveMode::Growth => "GROWTH".to_string(),
        DirectiveMode::ExploreStarved => "EXPLORE_STARVED".to_string(),
        DirectiveMode::MigrateCand => "MIGRATE_CAND".to_string(),
    }
}
